nncf/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,631 @@
+                        allow_nested_blocks: bool = True,
+                        allow_linear_combination: bool = False) -> List[BuildingBlock]:
+
+    """
+    This algorithm finds building blocks based on the analysis of the transformed graph.
+    A building block is a block that satisfies the following rules:
+    - has one input and one output tensors
+    - input and output tensors shape are the same
+    - removing a block from the graph (that is, the layers included in the block are not executed)
+      does not lead to duplication of edges along which the same tensor flows
+    - removing a block from the graph (that is, the layers included in the block are not executed)
+      does not lead to dangling edges
+    - removing a block from the graph (that is, the layers included in the block are not executed)
+      does not lead to duplicate activation layers
+    """
+
+    orig_graph = compressed_model.get_original_graph() # PTNNCFGraph
+    sgraph = prepare_search_graph(orig_graph)
+
+    fn_rules = [check_not_lead_to_double_edge,
+                check_not_lead_to_duplication_of_activation,
+                check_after_removing_block_graph_has_no_hanging_edges]
+
+    blocks = []
+    act_input_shape, act_output_shape = get_potential_candidate_for_block(sgraph)
+
+    for shape, start_nodes in act_input_shape.items():
+        for start_node in start_nodes:
+            pred_start_node = sgraph.get_prev_nodes(start_node.node_key)
+            if start_node.node_type == IGNORED_NAME_OPERATORS or len(pred_start_node) != 1:
+                continue
+            for end_node in act_output_shape[shape]:
+                if end_node.main_id - start_node.main_id > max_block_size:
+                    continue
+                if end_node.node_type in IGNORED_NAME_OPERATORS:
+                    continue
+                if end_node.main_id <= start_node.main_id:
+                    continue
+                is_one_edge = (end_node.main_id - start_node.bottom_id) == 1
+
+                # CHECK RULES
+                all_rules_is_true = True
+                for rule_fn in fn_rules:
+                    if not rule_fn(sgraph, start_node, end_node):
+                        all_rules_is_true = False
+                        break
+                if all_rules_is_true and not is_one_edge:
+                    blocks.append(BuildingBlock(start_node, end_node))
+
+    sorted_blocks = sorted(blocks, key=cmp_to_key(compare_for_building_block))
+    if not allow_linear_combination:
+        sorted_blocks = remove_linear_combination(sorted_blocks)
+    if not allow_nested_blocks:
+        sorted_blocks = remove_nested_blocks(sorted_blocks)
+    building_blocks_in_orig_graph = restore_node_name_in_orig_graph(sorted_blocks, orig_graph)
+
+    return building_blocks_in_orig_graph
+
+
+def remove_nested_blocks(sorted_blocks: List[BuildingBlock]) -> List[BuildingBlock]:
+    """
+    Remove nested building blocks.
```

Comment:
Return value explanation missing in docstring.

Reply:
Done.